---
title: "A2snp_H.lataniae_tolerance_Association"
author: "Casey_Flay"
date: "10/03/2021"
output:
  pdf_document: default
  html_document: default
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(data.table)
library(tidyverse)
module("load samtools/1.9 htslib/1.9 ncbi-blast/2.6.0")
```

##recode data and run kruskal.test for association of SNPs from A2 family with H.lataniae tolerance phenotype with SNPs filtered on heterozygous alleles in the tolerant Male parent. Significant sites +200bp were cut from the aligned reference PS1.68.5 and blast against PS1.69.0 to have a consistant genome alignments between datasets.
```{r recode data and kruskal.test}
A2all_long     <- fread("KWT_12E_run1_PS1.csv")
A2all_long     <- setnames(A2all_long,  c(names(A2all_long)), c(str_replace_all(names(A2all_long), ":", "_") ))
A2all_long     <- setnames(A2all_long, c(names(A2all_long)), c(str_replace_all(names(A2all_long), "-", "_") ))
A2all_long_nuc <- A2all_long #keep a nucleotide version for checking

A2all_long[A2all_long == "A"] <- "1"; A2all_long[A2all_long == "C"] <- "2"; A2all_long[A2all_long == "T"] <- "3"; A2all_long[A2all_long == "G"] <- "4"; A2all_long[A2all_long == "N"] <- "5"; A2all_long[A2all_long == "K"] <- "6"; A2all_long[A2all_long == "R"] <- "7"; A2all_long[A2all_long == "Y"] <- "8"; A2all_long[A2all_long == "W"] <- "9"; A2all_long[A2all_long == "M"] <- "10"; A2all_long[A2all_long == "S"] <- "11"; A2all_long[A2all_long == "Sus"] <- "21"; A2all_long[A2all_long == "Tol"] <- "20" #recode to numeric for data checking and heterozygous filter for appropriate parent

A2all_long <- suppressWarnings(A2all_long[, lapply(.SD, as.numeric), by=plant]) #change column data class to numeric for numeric filter

Female <- A2all_long[CK51_06_MumA >= 6] #only include snps which have a heterozygous parent at that SNP site
Male <- A2all_long[T.16_01_01A_Dad2 >= 6] #only include snps which have a heterozygous parent at that SNP site
MandF <- A2all_long[T.16_01_01A_Dad2 >= 6 | CK51_06_MumA >= 6]
## Change sex below for heterozygous alleles in that parent. The A2 family male is the parent of most interest as it carries tolerance. Female is used as a check.
A2all_wide <- as.data.table(t(MandF), keep.rownames=TRUE) #transpose table to wide format for KWT
A2all_wide <- setnames(A2all_wide, as.character(A2all_wide[1]))# needed as character for KWT.

A2all_wide <- A2all_wide[4:.N]# take out parents and column names

KWTtest    <- lapply(A2all_wide[1:.N], function(x){
kruskal.test(x, g=(A2all_wide$phenotype))
})

KWTtest    <- as.data.table(KWTtest)
KWTtest    <- KWTtest[3]
KWTtest    <- melt.data.table(KWTtest, id.vars = 1)
KWTtestall <- KWTtest[, plant := NULL]
KWTtest0.01 <- KWTtestall[value<0.01] #remove those with a high p value
KWTtest0.001 <- KWTtest0.01[value<0.001] #remove those with a high p value
Kruskal_p  <- as.numeric(KWTtest0.001$value)
snpat001   <- select(A2all_wide, KWTtest0.001$variable)
snpat001   <- names(snpat001)

SNP_check  <- A2all_long_nuc[plant %in% snpat001, ]
goodSNPs   <- cbind(SNP_check, KWTtest0.001$value)

KWT001plot <- qplot(KWTtest0.001$variable, Kruskal_p, xlab="SNPposition", ylab ="Kruskal.test_p-value")
KWT001plot + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))# these positions are for Red5_PS1.68.5 see below for PS 1.69.0
```

```{r get a file ready for blast alignment}
s<- as.data.table(str_split_fixed(KWTtest0.001$variable, pattern = "_", n=2))
t<-cbind(KWTtest0.001,s)
t<- t[V1!="phenotype"]
t<- t[, V2 := as.numeric(V2)]
t<- t[, "right" := V2+200]
t<- t[, V3 := str_replace_all(V1, "S", "CHR")]
t<- t[, V3 := paste0(V3,":", V2, "-", right)]
fwrite(t[, .(V3)], file="SNPposFromA2KWT.txt", col.names=FALSE)
#t<- t[V1=="S10"]
```

```{bash regions from A2 families 0.001, Eval:FALSE}
samtools faidx "/output/genomic/plant/Actinidia/chinensis/CK51F3_01/Genome/Assembly/PS1/1.68.5/AllChromosomes/PS1.1.68.5.fasta" -r "SNPposFromA2KWT.txt" -o "regionsSNPposFromA2KWT.fasta"
```

```{r Eval:FALSE}
file.symlink("/workspace/ComparativeDataSources/Actinidia/chinensis/Red5_PS1_1.69.0/GCA_003024255.1_Red5_PS1_1.69.0/GCA_003024255.1_Red5_PS1_1.69.0_genomic.fna", "/powerplant/workspace/hrtcdf/github/A2snps/alignbetweengenomes/reflink")
file.symlink("/workspace/ComparativeDataSources/Actinidia/chinensis/Red5_PS1_1.69.0/GCA_003024255.1_Red5_PS1_1.69.0/GCA_003024255.1_Red5_PS1_1.69.0_genomic.fna.fai", "/powerplant/workspace/hrtcdf/github/A2snps/alignbetweengenomes/reflink")
```

```{bash make blast database for Red5_PS1_1.69.0, eval:FALSE}
#module "load ncbi-blast/2.6.0"
Dir="blast"
ref="/powerplant/workspace/hrtcdf/github/A2snps/alignbetweengenomes/reflink/GCA_003024255.1_Red5_PS1_1.69.0_genomic.fna"

cd "/powerplant/workspace/hrtcdf/github/A2snps/alignbetweengenomes"
if  [ ! -d  "logDir" ]; then mkdir "logDir" ; fi
logDir="/powerplant/workspace/hrtcdf/github/A2snps/alignbetweengenomes/logDir"

bsub -n 8 -o $logDir/$Dir.log -e $logDir/$Dir.err "makeblastdb -in $ref -input_type fasta -dbtype nucl -title Red5_PS1_1.69.0"
```

```{bash blast A2family KWT hits+200bp against PS1.69, eval:FALSE}
#module "load ncbi-blast/2.6.0"
Dir="blastout"
ref="/powerplant/workspace/hrtcdf/github/A2snps/alignbetweengenomes/reflink/GCA_003024255.1_Red5_PS1_1.69.0_genomic.fna"

cd "/powerplant/workspace/hrtcdf/github/A2snps"
if  [ ! -d  "logDir" ]; then mkdir "logDir" ; fi
logDir="/powerplant/workspace/hrtcdf/github/A2snps/logDir"

bsub -n 8 -o $logDir/$Dir.log -e $logDir/$Dir.err "blastn -db $ref -query regionsSNPposFromA2KWT.fasta -out regionsSNPposFromA2KWT.out -evalue 1e-6 -outfmt 6"
```

```{r pull relavant bits as data.table and bind to original to check}
x <- fread("regionsSNPposFromA2KWT.out")
y <- x[V3==100, .(V1, V2, V9, V10)]
z <- KWTtest0.001[variable!="phenotype"]
y <- cbind(y, z)
s <- as.data.table(str_split_fixed(y$V1, pattern = ":", n=2))
s <- setnames(s, c("V1","V2"), c("V11", "V12"))
y <- cbind(y, s)
s <- as.data.table(str_split_fixed(y$V12, pattern = "-", n=2))
s <- setnames(s, c("V1","V2"), c("V13", "V14"))
y <- cbind(y, s)
y <- y[, V15 := str_replace_all(V11, "CHR", "S")
     ][, "positionPS1.68.5" := paste0(V15,"_", V13)
     ][, "positionPS1.69.0" := paste0(V11,"_", V9)
     ][, "positionMb" := V9/1000000
     ][,  positionMb := as.numeric(formatC(y$positionMb,digits=2,format="f"))
     ][, "positionPS1.69.0mb" := paste0(V11,"_", positionMb)
     ][, positionPS1.69.0mb := str_replace_all(positionPS1.69.0mb, "CHR", "Chr")
     ][, value := as.numeric(value)
     #][V11=="CHR10"
     ][order(V11,V9)]
sapply(y,class)
```

```{r}
options(repr.plot.width = 1, repr.plot.height = 0.05)
p <- y %>%
   
   ggplot(aes(x=positionPS1.69.0mb, y=y$value))+ 
   geom_jitter(stat="identity", width = 0)+
   labs(y= "Kruskal.test_p-value", x = "SNPposition on Chromosome 10")
p <- p + theme(axis.text.x = element_text(angle = 90))
#p <- p + scale_x_continuous(name="Mbp along chromosome 10", limits=c(0, 20))
ggsave("A2kwt.png", plot = p)
p
```
```{r}
p <- y %>%
   ggplot( aes(x=positionMb, y=y$value))+ 
   geom_point(stat="identity")+
   labs(y= "Kruskal.test_p-value", x = "SNPposition on Chromosome 10")
p <- p + scale_x_continuous(name="Mbp along chromosome 10", limits=c(0, 20))
ggsave("A2kwt.png", plot = p)
p
```


