---
title: "A2snp_H.lataniae_tolerance_Association"
author: "Casey_Flay"
date: "10/03/2021"
output:
  pdf_document: default
  html_document: default
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(data.table)
library(tidyverse)
```

```{r recode data and kruskal.test for association of SNPs with H.lataniae tolerance phenotypefrom the A2 population with SNPs filtered on heterozygous alleles in the Female or Male parent}
A2all_long     <- fread("KWT_12E_run1_PS1.csv")
A2all_long     <- setnames(A2all_long,  c(names(A2all_long)), c(str_replace_all(names(A2all_long), ":", "_") ))
A2all_long     <- setnames(A2all_long, c(names(A2all_long)), c(str_replace_all(names(A2all_long), "-", "_") ))
A2all_long_nuc <- A2all_long #keep a nucleotide version for checking

A2all_long[A2all_long == "A"] <- "1"; A2all_long[A2all_long == "C"] <- "2"; A2all_long[A2all_long == "T"] <- "3"; A2all_long[A2all_long == "G"] <- "4"; A2all_long[A2all_long == "N"] <- "5"; A2all_long[A2all_long == "K"] <- "6"; A2all_long[A2all_long == "R"] <- "7"; A2all_long[A2all_long == "Y"] <- "8"; A2all_long[A2all_long == "W"] <- "9"; A2all_long[A2all_long == "M"] <- "10"; A2all_long[A2all_long == "S"] <- "11"; A2all_long[A2all_long == "Sus"] <- "21"; A2all_long[A2all_long == "Tol"] <- "20" #recode to numeric for data checking and heterozygous filter for appropriate parent

A2all_long <- suppressWarnings(A2all_long[, lapply(.SD, as.numeric), by=plant]) #change column data class to numeric for numeric filter

Female     <- "CK51_06_MumA"
Male       <- "T.16_01_01A_Dad2"


## Change sex below for heterozygous alleles in that parent. The male is the parent of most interest as it carries tolerance. Female is used as a check.
A2all_long <- A2all_long[Male >= 6] #only include snps which have a heterozygous parent at that SNP site

A2all_wide <- as.data.table(t(A2all_long), keep.rownames=TRUE) #transpose table to wide format for KWT
A2all_wide <- setnames(A2all_wide, as.character(A2all_wide[1]))# needed as character for KWT.

A2all_wide <- A2all_wide[4:.N]# take out parents and column names

KWTtest    <- lapply(A2all_wide[1:.N], function(x){
kruskal.test(x, g=(A2all_wide$phenotype))
})

KWTtest    <- as.data.table(KWTtest)
KWTtest    <- KWTtest[3]
KWTtest    <- melt.data.table(KWTtest, id.vars = 1)
KWTtest    <- KWTtest[, plant := NULL]
KWTtest001 <- KWTtest[value<0.001] #remove those with a high p value
Kruskal_p  <- as.numeric(KWTtest001$value)
snpat001   <- select(A2all_wide, KWTtest001$variable)
snpat001   <- names(snpat001)

SNP_check  <- A2all_long_nuc[plant %in% snpat001, ]
goodSNPs   <- cbind(SNP_check, KWTtest001$value)

KWT001plot <- qplot(KWTtest001$variable, Kruskal_p, xlab="SNPposition", ylab ="Kruskal.test_p-value")
KWT001plot + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

```{r kruskal.test for association of genotypes from the A1 and A2 population with H.lataniae tolerance phenotype}
A1 <- fread("A1popKWT.csv")
A1r <- kruskal.test(A1$KWTgenotype1234, g=(A1$KWTphenotype))
print(A1r)

A2 <- fread("A2popKWT.csv")
A2r <- kruskal.test(A2$KWTgenotype1234, g=(A2$KWTphenotype))
print(A2r)
```
